File: packages/dart_super/lib/src/rx/rx_notifier.dart
part of 'rx.dart';

/// {@template rx_notifier}
/// An abstract base class for creating reactive notifiers that manage
/// a state of type `T`.
///
/// The `RxNotifier` class provides a foundation for creating reactive notifiers
/// that encapsulate a piece of immutable state and notify their listeners
/// when the state changes. Subclasses of `RxNotifier` must override the
/// `watch` method to provide the initial state and implement the logic
/// for updating the state.
///
/// Example usage:
///
/// ```dart
/// class CounterNotifier extends RxNotifier<int> {
///   @override
///   int watch() {
///     return 0; // Initial state
///   }
///
///   void increment() {
///     state++; // Update the state
///   }
/// }
///
/// final counter = CounterNotifier();
///
/// // Adding a listener to the notifier
/// counter.addListener(() {
///   print('Counter changed: ${counter.state}');
/// });
///
/// // Updating the state
/// counter.increment(); // This will trigger the listener and print the updated state.
/// ```
///
/// It is best used for global state i.e state used in multiple controllers
/// but it could also be used for a single controller to abstract a
/// state and its events e.g if a state has a lot events, rather than
/// complicating your controller, you could use an RxNotifier for that singular
/// state instead.
///
/// **Note:** When using the RxNotifier class, it is important to call the
/// `dispose()` method on the object when it is no longer needed to
/// prevent memory leaks.
/// This can be done using the onDisable method of your controller.
/// {@endtemplate}
abstract class RxNotifier<T> extends Rx<T> {
  /// {@macro rx_notifier}
  RxNotifier() {
    _state = watch();
  }

  late T _state;
  bool _loading = false;

  /// Retrieves the initial state for the notifier.
  ///
  /// Subclasses must override this method to provide the initial state for
  /// the notifier. The initial state is returned and used as the current state
  /// when the notifier is created.
  ///
  /// Example:
  ///
  /// ```dart
  /// @override
  /// int watch() {
  ///   return 0; // Initial state
  /// }
  /// ```
  @protected
  T watch();

  /// The current state of the notifier.
  T get state {
    RxListener._read(this);
    return _state;
  }

  /// The loading state of the notifier.
  ///
  /// This is useful when working with asynchronous state.
  /// By default it is set to false so that none asynchronous state
  /// can be utilized.
  ///
  /// Example:
  ///
  /// ```dart
  /// class BooksNotifier extends RxNotifier<List<Book>> {
  ///   @override
  ///   List<Book> watch() {
  ///     return []; // Initial state
  ///   }
  ///
  ///   void getBooks() async {
  ///     toggleLoading();
  ///     state = await booksRepo.getBooks(); // Update the state
  ///   }
  /// }
  /// ```
  bool get loading => _loading;

  /// Sets the new state for the notifier.
  ///
  /// The `state` argument represents the new state value for the notifier.
  /// If the new state is different from the current state, the notifier will
  /// update its state and notify its listeners.
  ///
  /// Example:
  ///
  /// ```dart
  /// void increment() {
  ///   state++; // Update the state
  /// }
  /// ```
  @protected
  @visibleForTesting
  set state(T state) {
    if ('$_state' == '$state') return;
    if (loading == true) toggleLoading();
    _state = state;
    _notifyListeners();
  }

  /// Toggles the loading state of the notifier. By default the
  /// loading state is false.
  ///
  /// Example:
  ///
  /// ```dart
  /// void getList() async {
  ///   toggleLoading(); // set loading to true
  ///
  ///   state = await dataRepo.getData; // Update the state
  /// }
  /// ```
  /// **Note**: There is no need to call it a second time, when the state
  /// is updated the loading state will be set to false.
  @protected
  @visibleForTesting
  void toggleLoading() => _loading = !_loading;

  @override
  String toString() => '$runtimeType($state)';
}

File: packages/js-super/src/rx/rx_notifier.ts
import { Rx, RxListener, VoidCallback } from "./rx";

/**
 * Represents a reactive object that holds a state value of type `T`,
 * provides a watch function to track changes, and notifies listeners of state changes.
 * @template T The type of the state value.
 */
abstract class RxNotifier<T extends Object> extends Rx<T> {
    /**
     * Creates a new `RxNotifier` object.
     * The state value is initialized with the result of the `watch` function.
     */
    constructor() {
      super();
      this._state = this.watch();
    }
    
    private _listeners: Array<VoidCallback> = [];
    private _mounted = true;
    private _state: T;
    
    /**
     * Abstract method to be implemented by subclasses.
     * The `watch` function should return the current state value.
     * @returns The current state value.
     */
    abstract watch(): T;  
    
    get state(): T {
      RxListener.read(this);
      return this._state;
    }
  
    set state(value: T) {
      if (`${this._state}` === `${value}`) return;
      this._state = value;
      this._notifyListeners();
    }
    
    get mounted(): boolean {
      return this._mounted;
    }
  
    get hasListeners(): boolean {
      return this._listeners.length > 0;
    }
  
    addListener(listener: VoidCallback): void {
      this._listeners.push(listener);
    }
  
    removeListener(listener: VoidCallback): void {
      this.assertMounted();
      const index = this._listeners.indexOf(listener);
      if (index !== -1) {
        this._listeners.splice(index, 1);
      }
    }
  
    dispose(): void {
      if (!this._mounted) return;
      this._mounted = false;
      this._listeners = [];
    }
  
    /**
     * Notifies registered listeners about changes in the object's state.
     * This method should be called whenever the object's state changes.
     * It should not be called after the object has been disposed.
     */
    private _notifyListeners(): void {
      this.assertMounted();
      for (const listener of this._listeners) {
        listener();
      }
    }
}

export default RxNotifier;
